cmd/tier-sweeper.go

```diff
@@ -0,0 +1,155 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"net/http"
+	"strings"
+
+	xhttp "github.com/minio/minio/cmd/http"
+	"github.com/minio/minio/pkg/bucket/lifecycle"
+)
+
+// objSweeper determines if a transitioned object needs to be removed from the remote tier.
+// A typical usage would be like,
+// os := newObjSweeper(bucket, object)
+// // Perform a ObjectLayer.GetObjectInfo to fetch object version information
+// goiOpts := os.GetOpts()
+// gerr := objAPI.GetObjectInfo(ctx, bucket, object, goiOpts)
+// if gerr == nil {
+//    os.SetTransitionState(goi)
+// }
+//
+// // After the overwriting object operation is complete.
+// if jentry, ok := os.ShouldRemoveRemoteObject(); ok {
+//     err := globalTierJournal.AddEntry(jentry)
+//     logger.LogIf(ctx, err)
+// }
+type objSweeper struct {
+	Object           string
+	Bucket           string
+	ReqVersion       string // version ID set by application, applies only to DeleteObject and DeleteObjects APIs
+	Versioned        bool
+	Suspended        bool
+	TransitionStatus string
+	TransitionTier   string
+	RemoteObject     string
+}
+
+// newObjSweeper returns an objSweeper for a given bucket and object.
+// It initializes the versioning information using bucket name.
+func newObjSweeper(bucket, object string) *objSweeper {
+	versioned := globalBucketVersioningSys.Enabled(bucket)
+	suspended := globalBucketVersioningSys.Suspended(bucket)
+	return &objSweeper{
+		Object:    object,
+		Bucket:    bucket,
+		Versioned: versioned,
+		Suspended: suspended,
+	}
+}
+
+// versionID interface is used to fetch object versionID from disparate sources
+// like http.Request and ObjectToDelete.
+type versionID interface {
+	GetVersionID() string
+}
+
+// multiDelete is a type alias for ObjectToDelete to implement versionID
+// interface
+type multiDelete ObjectToDelete
+
+// GetVersionID returns object version of an object to be deleted via
+// multi-delete API.
+func (md multiDelete) GetVersionID() string {
+	return md.VersionID
+}
+
+// singleDelete is a type alias for http.Request to implement versionID
+// interface
+type singleDelete http.Request
+
+// GetVersionID returns object version of an object to be deleted via (simple)
+// delete API. Note only when the versionID is set explicitly by the application
+// will we return a non-empty versionID.
+func (sd singleDelete) GetVersionID() string {
+	return strings.TrimSpace(sd.URL.Query().Get(xhttp.VersionID))
+}
+
+// ForDelete sets the version ID from v
+func (os *objSweeper) ForDelete(v versionID) *objSweeper {
```

Comment:
`ForDelete` doesn't tell much. Maybe:

```suggestion
func (os *objSweeper) DeleteVersion(v versionID) *objSweeper {
```
or if it makes sense:

```suggestion
func (os *objSweeper) WithVersion(v versionID) *objSweeper {
```